### ISSUE TITLE
Path Coverage enhancement

### ISSUE THREAD

**Issue by CuiYifeng** [intel/torch-xpu-ops#2127]
### 🚀 The feature, motivation and pitch

In #2073, the root cause was located in a code path that has never been executed before. We need to consider how to avoid similar issues.

**Comment by CuiYifeng** [comment]
@laifenxiawucha triage

**Comment by laifenxiawucha** [comment]
Review requested in [this mention](https://github.com/intel/torch-xpu-ops/issues/2127#issuecomment-4264786723).

Now I have enough context to provide a helpful triage. This is asking about path coverage enhancement to avoid bugs in untested code paths, but the issue is quite minimal. I need to ask for clarification to understand what specific solution they're looking for.

---

## Triage Summary

**Needs more information**

## Next Step

To provide a concrete recommendation about improving path coverage, we need to understand the scope and goal more clearly. Please provide:

1. **Brief context about #2073**: What type of code path was untested? (e.g., a specific kernel variant, a dispatch branch, error handling path, or Python wrapper edge case?)

2. **Expected outcome**: Are you looking for one of the following, or something else?
   - Code coverage tools integrated into CI (e.g., pytest-cov, coverage.py)
   - Targeted test expansion for specific operator families
   - Automated detection of uncovered dispatch paths
   - A policy or process change for new PRs

3. **Meta vs. actionable goal**: Is this issue intended as a planning discussion to gather ideas, or are you ready to propose a specific approach and begin implementation?

With these details, we can identify the right subsystem (testing infrastructure, CI configuration, or operator-specific test coverage) and provide a concrete path forward.

**Comment by laifenxiawucha** [comment]
## Triage

Valid concern. PR #2073 fixed an out-of-bounds access in BatchLinearAlgebraOps error handling that existed in an untested code path.

To prevent similar issues, consider:

1. **Fuzz testing** for error paths — generate random invalid inputs to exercise error handling
2. **Static analysis** — tools like clang-tidy or Coverity can detect out-of-bounds access patterns in functions that only execute on error paths
3. **Unit tests for error conditions** — explicitly test `TORCH_CHECK` and exception scenarios rather than just happy paths
4. **Coverage tool integration** — enforce minimum branch coverage (not just line coverage) for critical ops

Suggested next step: Add a `fuzz` category to test directory with error-condition tests for linear algebra ops.

*[AI-assisted reply]*

**Comment by CuiYifeng** [comment]
@laifenxiawucha issue-processing

**Comment by laifenxiawucha** [comment]
## Issue Processing

Triage complete. To proceed with the suggested next step, create the following structure:

**1. Create test directory**
```
tests/xpu/fuzz/
├── __init__.py
├── linear_algebra/
│   ├── __init__.py
│   └── test_error_paths.py
└── conftest.py
```

**2. Write error-path tests for BatchLinearAlgebraOps**
```python
# tests/xpu/fuzz/linear_algebra/test_error_paths.py
import pytest
import torch
import torch.xpu

@pytest.mark.fuzz
def test_batched_linalg_invalid_dims():
    # Test out-of-bounds access trigger from #2073
    with pytest.raises(RuntimeError):
        x = torch.randn(2, 2).xpu()
        torch.linalg.lu(x)  # Wrong shape path
```

**3. Add to CI workflow**
Extend `.github/workflows/test.yml` with Fuzz test matrix.

**Labels to add**: [`testing`, `enhancement`]

**Sub-task issue**: "Add fuzz test category for linear algebra error paths" if further breakdown needed.

*[AI-assisted reply]*

**Comment by laifenxiawucha** [comment]
Review requested in [this mention](https://github.com/intel/torch-xpu-ops/issues/2127#issuecomment-4264842584).

### Triage: #2127 — Path Coverage enhancement

**Level**: L3

**Case Summary**: This is a feature enhancement request (0 failing cases / 0 total).

**Analysis**:

This issue is a meta-level infrastructure investigation. The context is clear: PR #2073 exposed a latent bug in error handling (`src/ATen/native/xpu/mkl/BatchLinearAlgebra.cpp:57`) that lived in a code path never executed by tests. The original loop iterated over an empty `ids` vector, so the error-rethrow body never ran and the out-of-bounds issue remained undetected. The fix changed iteration to `errs` directly, enabling the error path to be exercised safely. Line coverage alone cannot surface such unreachable/untried branches; path/branch coverage is needed.

**Next Steps**:

- Confirm practical options and data formats: Many coverage tools emit branch/path information (e.g., gcc/clang’s `--coverage` `.gcda` reports include branch details; some static analyzers can flag unexecuted code). Identify whether branch/path coverage targets can be generated in this build system and whether they already exist or need new steps added.
- Assess integration trade-offs and platform constraints: Evaluate tooling availability on supported platforms, cost (build time, runtime overhead, storage), and whether integration aligns with existing CI stages.
- Provide actionable guidance: If immediate integration is infeasible, document quick wins (e.g., code patterns to write testable error paths, optional local/use-time branch coverage) and outline a roadmap for CI integration.

**Comment by CuiYifeng** [comment]
@laifenxiawucha triage

**Comment by laifenxiawucha** [comment]
Review requested in [this mention](https://github.com/intel/torch-xpu-ops/issues/2127#issuecomment-4264919169).

Now I understand the context. PR #2073 fixed an out-of-bounds access in error handling for BatchLinearAlgebra Ops - a bug that lived in a rarely executed code path (exception rethrowing during MKL failures). Issue #2127 is asking how to prevent similar issues from reaching production by improving path coverage.

Here's my triage:

**📋 Triage: Enhancement - Test coverage improvement**

**Summary:** This issue proposes improving path coverage after discovering a bug in a rarely executed error-handling code path. The request is valid but the scope and approach need clarification before implementation.

**Next step:** The issue needs more specification on what type of path coverage improvement is desired.

**Missing information:**
1. What specific coverage metric or tool should we target (e.g., code coverage percentage for error paths, fuzzing integration, static analysis)?
2. Should this focus specifically on error-handling paths in MKL/oneAPI集成, or be a general coverage policy?
3. Is the goal a CI enforcement mechanism (e.g., block PRs with low coverage), an investigative tool for reviewers, or a specific test framework addition?
4. Are there existing coverage tools or processes in the repo to build upon, or is this starting from scratch?

**Recommendation:** Once the above questions are answered, consider focusing on error-injection testing for MKL operations (`BatchLinearAlgebra.cpp`, other MKL wrappers) since those contain complex error-handling paths that are hard to trigger naturally in unit tests.

**Comment by laifenxiawucha** [comment]
## Triage Summary

Valid enhancement concern. The issue in #2073 (out-of-bounds access in error handling) was in an untested code path.

### Recommended Actions

1. **Add fuzz testing** for exception/error paths, especially for operations with complex error handling (e.g., BatchLinearAlgebra)
2. **Covered-by correctness**: For each new op, verify all TORCH_CHECK branches have corresponding test cases that trigger them (invalid inputs, device mismatch, unsupported dtypes)
3. **CI gap analysis**: Use code coverage tools (llvm-cov, gcov) to identify unTested paths and prioritize adding tests for them
4. **Error path testing**: Structural fuzzer that randomly mocks invalid internal states (out-of-bounds indices, null pointers) can catch memory safety issues in error handling

### Priority

Medium enhancement - good for reducing latent bugs but doesn't block current feature work.

*[AI-assisted reply]*

**Comment by CuiYifeng** [comment]
@laifenxiawucha How can I prioritize you to execute `triage` in a graded manner like https://github.com/intel/torch-xpu-ops/issues/2127#issuecomment-4264856212?

**Comment by laifenxiawucha** [comment]
Review requested in [this mention](https://github.com/intel/torch-xpu-ops/issues/2127#issuecomment-4264979803).

Thanks for the feedback about prioritizing triage execution. I've noted your request for graded triage prioritization and queued it for the weekly bot-review batch to improve routing clarity.

For the current #2127 issue about path coverage enhancement, if you'd like me to perform an L1/L2/L3 structural issue triage now, just ask and I'll analyze this issue's actionability, missing information, and next steps.

**Comment by CuiYifeng** [comment]
@laifenxiawucha triage

**Comment by laifenxiawucha** [comment]
Review requested in [this mention](https://github.com/intel/torch-xpu-ops/issues/2127#issuecomment-4265009453).

### Triage: **Enhancement — Path Coverage Enhancement**

**Level**: Enhancement

**Analysis**: This is a process/test infrastructure enhancement request. The issue correctly identifies a class of bugs: untested error paths (such as error handling in ops like `lu_factor`) that remain dormant until triggered in production. You already have `skip_list` tracking and a `skipped_ut` runner, so the question is what additional coverage strategy or tooling you want.

**Next Steps**:
To help scope the solution, please specify which approach (or combination) you're considering:

1. **CI Job**: Add a periodic "run all skipped tests" job with path coverage instrumentation (e.g., pytest-cov) to report which code lines remain unexecuted across the full test surface.
2. **Kernel-Level Coverage**: Instrument the XPU kernel paths (symmetric to DispatchStub kernel detection) and require coverage gating for new error-handling code.
3. **Static Analysis**: Add coverage-aware static analysis to flag error/exception paths without corresponding tests (similar to how pytest-cov reports uncovered lines).
4. **Audit Workflow**: Define a one-time audit process to enumerate all `catch`, error, and fallback paths in `torch-xpu-ops` and incrementally add targeted regression tests.

If you have a draft proposal or examples from other backends (PyTorch CUDA/CPU), linking those would help narrow the scope.

**Comment by laifenxiawucha** [comment]
> [@laifenxiawucha](https://github.com/laifenxiawucha) How can I prioritize you to execute `triage` in a graded manner like [#2127 (comment)](https://github.com/intel/torch-xpu-ops/issues/2127#issuecomment-4264856212)?我该如何优先安排您以分级方式执行 `triage` 例如 [#2127 (comment)](https://github.com/intel/torch-xpu-ops/issues/2127#issuecomment-4264856212) ？

This accidentally set off the feedback skill. It’s not quite ready yet, so I’ve turned it off for now.

**Comment by laifenxiawucha** [comment]
Replying to [this mention](https://github.com/intel/torch-xpu-ops/issues/2127#issuecomment-4265229114).

I see you're asking how to trigger a graded L1/L2/L3 triage for an issue. Here's the key:

To get me to perform a graded triage, you need to:

1. @mention me in the issue thread (or comment thread on a PR/commit)
2. Explicitly use the word `triage` in your request
3. Specify which issue you want triaged by number or URL (e.g., "triage #2127" or "triage https://github.com/intel/torch-xpu-ops/issues/2127")

For example, a working request would be:
```
@mention triage #2127
```

Once you do that, I'll analyze the issue and classify it as L1 (clear/trivial), L2 (moderate investigation needed), or L3 (deep analysis/complexity), then provide level-specific analysis and next steps.

Would you like me to perform that triage on this issue (#2127) now?
